pkg/featurechangestopper/featurechangestopper.go

```diff
@@ -0,0 +1,140 @@
+// or from anything to TechPreviewNoUpgrade.
+type TechPreviewChangeStopper struct {
+	startingTechPreviewState bool
+
+	featureGateLister configlistersv1.FeatureGateLister
+	cacheSynced       []cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
+}
+
+// New returns a new cluster version operator.
+func New(
+	startingTechPreviewState bool,
+	featureGateInformer configinformersv1.FeatureGateInformer,
+) *TechPreviewChangeStopper {
+	c := &TechPreviewChangeStopper{
+		startingTechPreviewState: startingTechPreviewState,
+		featureGateLister:        featureGateInformer.Lister(),
+		cacheSynced:              []cache.InformerSynced{featureGateInformer.Informer().HasSynced},
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "feature-gate-stopper"),
+	}
+
+	featureGateInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.queue.Add("cluster")
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			c.queue.Add("cluster")
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.queue.Add("cluster")
+		},
+	})
+
+	return c
+}
+
+func (c *TechPreviewChangeStopper) syncHandler(ctx context.Context, shutdownFn context.CancelFunc) error {
+	featureGates, err := c.featureGateLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+	techPreviewNowSet := featureGates.Spec.FeatureSet == configv1.TechPreviewNoUpgrade
+	if techPreviewNowSet != c.startingTechPreviewState {
+		klog.Infof("TechPreviewNoUpgrade status changed from %v to %v, shutting down.", c.startingTechPreviewState, techPreviewNowSet)
+		if shutdownFn != nil {
+			shutdownFn()
+		}
+		return nil
+	}
+	return nil
+}
+
+func (c *TechPreviewChangeStopper) Run(runContext context.Context, shutdownFn context.CancelFunc) {
+	// don't let panics crash the process
+	defer utilruntime.HandleCrash()
+	// make sure the work queue is shutdown which will trigger workers to end
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting stop-on-techpreview-change controller")
+
+	// wait for your secondary caches to fill before starting your work
+	if !cache.WaitForCacheSync(runContext.Done(), c.cacheSynced...) {
+		return
+	}
+
+	// runWorker will loop until "something bad" happens.  The .Until will
+	// then rekick the worker after one second
+	go wait.UntilWithContext(runContext, func(ctx context.Context) {
+		c.runWorker(ctx, shutdownFn)
+	}, time.Second)
+
+	// wait until we're told to stop
+	<-runContext.Done()
+	klog.Infof("Shutting down stop-on-techpreview-change controller")
+}
+
+func (c *TechPreviewChangeStopper) runWorker(ctx context.Context, shutdownFn context.CancelFunc) {
+	// hot loop until we're told to stop.  processNextWorkItem will
+	// automatically wait until there's work available, so we don't worry
+	// about secondary waits
+	for c.processNextWorkItem(ctx, shutdownFn) {
```

Comment:
nit: There's not much going on in `runWorker` besides the comment here.  `Run` itself is fairly short and suite, so I dunno if it's worth a separate shim function just to get this comment out of `Run`.  Can we drop `runWorker` and inline its contents in `Run`?

Reply:
> nit: There's not much going on in runWorker besides the comment here. Run itself is fairly short and sweet, so I dunno if it's worth a separate shim function just to get this comment out of Run. Can we drop runWorker and inline its contents in Run?


This is the standard structure from https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md#rough-structure  .  I suggest keeping it the same to match standard kube controllers.

Reply:
I've floated kubernetes/community/pull/6267 to improve shutdown/irrecoverable-error handling in the upstream template.  Can you review and consider picking up any portions that make sense to you?

Reply:
Deferring this change; I'll file a follow-up CVO PR to pick up the stuff I'm suggesting in kubernetes/community#6267.